containers/daap-python-base/src/var/task/data_platform_paths.py

```diff
@@ -0,0 +1,259 @@
+"""
+Utilities for constructing and parsing S3 paths for a data product,
+and the corresponding athena tables.
+
+Example for data product name "data_product", table name "table":
+
+- Raw data is stored at: raw_data/data_product/table/extraction_timestamp=timestamp/file.csv
+- Curated data is stored at:
+  curated_data/database_name=data_product/table_name=table/extraction_timestamp=timestamp/file.parquet
+- Athena table name for raw data is: data_products_raw.table_raw
+- Athena table name for curated data is: data_product.table
+"""
+
+from __future__ import annotations
+
+import os
+import re
+from datetime import datetime
+from typing import NamedTuple
+from uuid import UUID
+
+import boto3
+
+RAW_DATABASE_NAME = "data_products_raw"
+EXTRACTION_TIMESTAMP_FORMAT = "%Y%m%dT%H%M%SZ"
+EXTRACTION_TIMESTAMP_REGEX = re.compile(
+    r"^(.*)/(extraction_timestamp=)([0-9TZ]{1,16})/(.*)$"
+)
+
+
+class BucketPath(NamedTuple):
+    """
+    A path to an object in S3
+    """
+
+    bucket: str
+    key: str
+
+    @property
+    def uri(self):
+        return f"s3://{self.bucket}/{self.key}"
+
+    @property
+    def parent(self):
+        return os.path.dirname(self.key)
+
+    @staticmethod
+    def from_uri(uri):
+        if not uri.startswith("s3://"):
+            raise ValueError(uri)
+
+        bucket, key = uri.replace("s3://", "").split("/", 1)
+
+        return BucketPath(bucket, key)
+
+
+class QueryTable(NamedTuple):
+    """
+    Identifies a qualified table in AWS
+    """
+
+    database: str
+    name: str
+
+
+def get_bucket_name() -> str:
+    """
+    Get the bucket name from the environment
+    """
+    return os.environ["BUCKET_NAME"]
+
+
+def get_account_id() -> str:
+    """
+    Get the account ID from the environment / AWS configuration
+    """
+    return boto3.client("sts").get_caller_identity()["Account"]
+
+
+class DataProductConfig:
+    """
+    Configures the name, S3 paths, and athena table names for a data product.
+    """
+
+    def __init__(self, name: str, table_name: str, bucket_name: str | None = None):
+        """
+        Generate all the paths based on the data product name and a table name
+        """
+        if bucket_name is None:
+            bucket_name = get_bucket_name()
+
+        self.name = name
+
+        self.raw_data_prefix = BucketPath(
+            bucket=bucket_name, key=os.path.join("raw_data", name, table_name) + "/"
+        )
+
+        self.curated_data_prefix = BucketPath(
+            bucket_name,
+            os.path.join(
+                "curated_data",
+                f"database_name={name}",
+                f"table_name={table_name}",
+            )
+            + "/",
+        )
+
+        self.raw_data_table = QueryTable(
+            database=RAW_DATABASE_NAME, name=f"{table_name}_raw"
```

Comment:
I noticed this table name does not contain the data product name. Could this cause problems if the table names clash?

(I based this on `containers/daap-athena-load/src/var/task/create_curated_athena_table.py`)

Reply:
this table is temporary, and is deleted on every run. Saying that, there is an issue (https://github.com/ministryofjustice/data-platform/issues/1310) to add some random string to the name of it as it could cause potential issues with quick consecutive loads of the same data product